dist/stat_data.py

```diff
@@ -335,6 +339,41 @@ def __init__(self, name, desc, flags=''):
     CapacityStat('fsync_all_fh_total', 'background fsync file handles considered'),
     CapacityStat('fsync_all_time', 'background fsync time (msecs)', 'no_clear,no_scale'),
 
+    ##########################################
+    # Checkpoint statistics
+    ##########################################
+    CheckpointStat('checkpoints', 'transaction checkpoints'),
+    CheckpointStat('checkpoint_custom_dsrc', 'number of custom data sources checkpointed'),
+    CheckpointStat('checkpoint_fsync_post', 'transaction fsync calls for checkpoint after allocating the transaction ID'),
+    CheckpointStat('checkpoint_fsync_post_duration', 'transaction fsync duration for checkpoint after allocating the transaction ID (usecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_generation', 'transaction checkpoint generation', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_handle_applied', 'transaction checkpoint most recent handles applied'),
+    CheckpointStat('checkpoint_handle_checked', 'transaction checkpoint data handle operation applied'),
```

Comment:
This description is not clear to me. What is the difference between the most recent number of data handles applied (the line above) and simply "data handle operation applied"? I don't know what a data handle operation means in the context of checkpoint (or outside of it).

Reply:
In the current tree for stats, `handles_walked = handles_skipped + handles_applied`.

Reply:
This is basically counting all of the dhandles that `__checkpoint_apply_to_dhandles` looks at. As far as I can tell, none of these visits were being tracked, even though they were hiding potentially expensive operations (e.g. `__wt_checkpoint_sync`). I guess the alternative would be a counter for each of the functions passed in to `__checkpoint_apply_to_dhandles`

Reply:
So `checkpoint_apply_to_dhandles` is called in several locations. It seems like this new stat will be just `handles_applied` times number of calls to `checkpoint_apply_to_dhandles` for the various operations.

Reply:
OK, I'd missed that `checkpoint_handle_applied` was also counting this. Looking at T2 again, I think there's some value in using the new stat instead since it will increment as it goes, rather than in one big increase at the end.

The way the existing stat is implemented raises a performance question: are we willing to pay the cost of having these stats incremented as we go? I think per-dhandle is OK, but I'm less confident about the per-page stats I just added. (CC-ing @agorrod)

Reply:
Sorry for the delay Will - I missed the ping.

I think your assessment is right. A stat for visiting a dhandle is fine (though it should add value and be clearly understood). I'd be careful adding new statistics that increment for each reference a checkpoint visits (page in-or-out of cache) - one or two of those is probably fine, if they add value.

Adding some statistics that capture information related to when checkpoint is doing something meaningful to a page would be fine.

I'd rather approach this the other way though. We've seen some cases where checkpoints take a long time, and it isn't possible to see where that time is going. How can we get that insight - once we know that, we can figure out whether always getting that insight is too expensive to be worthwhile.

I've skimmed this PR, but am not yet confident that the additional information will give us the information. How confident are you about that?

Reply:
I've got a test scenario that causes quite long checkpoints - can't remember the exact timing since it was a few weeks ago. Most of the time was spent doing IO, which makes sense, but it's also not a scenario that would surface performance issues with per-page stats.

I'll spend some time trying to come up with a checkpoint scenario that isn't IO-bound, and see what's useful and what isn't.

Reply:
@agorrod I've done some more testing, I couldn't quite get it to not be IO-bound but it's still interesting. tl;dr is that I think we should get rid of `handles_walked`, `handles_skipped` and `handles_applied` in favour of the new stats (possible renaming the new ones for clarity), and the per-page stats could be bumped every (say) 1024 pages.

Reply:
Thanks for the context, I didn't know about that issue. Stepping back and thinking about this again with that in mind - there were two reasons I added "data handle operations applied". One was that it's possible to skip the operation, and that's not tracked anywhere:

```
for (i = 0; i < session->ckpt_handle_next; ++i) {
    if (session->ckpt_handle[i] == NULL)
        continue;
```

and the second reason was that we apply several functions to all dhandles, and that also wasn't visible (e.g. `__checkpoint_presync`).

What about:
1. Dropping the original "checkpoint handle applied" 
2. Renaming the new "data handle operations applied" to "checkpoint handle applied"
3. Changing "checkpoint_handle_skipped" and "checkpoint handle walked" to update incrementally rather than once we're done applying all of the ops?

Reply:
Can you describe how/why you want to update the skipped and walked incrementally? We determine those values when we gather the handles at the beginning of checkpoint. Then we update based on what we did when walking the handles.

Reply:
Sure - my concern is that either (or both) gathering the dhandles and walking the dhandles is what could be expensive. If that process is taking a long time, it's not clear from T2 (at the moment) that that's where the time is going. Whereas updating them incrementally would show (say) a large pause if an operation on a particular dhandle is expensive, or a gradual increase (instead of a large spike at the end) if the problem is the quantity of dhandles.

That said, checkpoint is something of a blind spot for me in the WT codebase - maybe these two steps being slow isn't realistic and my concern is misplaced.

Reply:
Fair enough. I'll drop this new "data handle operations applied" stat then, thanks for taking the time to explain the use case and background here.

Reply:
Done in ae54606